File: LeetCode/0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.swift
import Foundation

class Solution {
    func findPairs(_ nums: [Int], _ k: Int) -> Int {
        let sorted = nums.sorted()
        var left = 0, right = 1
        var ans = Set<Int>()

        while (right < nums.count) {
            let diff = sorted[right] - sorted[left]
            if diff == k {
                ans.insert(sorted[right] + sorted[left])
                left += 1
                right += 1
            } else if diff < k {
                right += 1
            } else { 
                left += 1
                right += left == right ? 1 : 0
            }
        }

        return ans.count
    }
}

File: LeetCode/0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findPairs = function(nums, k) {
    let count = new Set(), l = 0, r = 1;

    nums.sort((a,b) => a - b);

    while (r < nums.length) {
        let diff = nums[r] - nums[l];
        if (diff == k) {
            count.add(nums[l] + nums[r]);
            l++;
            r++;
        } else if (diff < k) {
            r++;
        } else {
            l++;
            r += (l == r) ? 1 : 0
        }
    }

    return [...count].length;
};